lib: lte_link_control: Add support for %CONEVAL AT command

simensrostad opened the pull request:
Add support for AT command that can be used to pre-evaluate the
LTE radio signal state prior to data transmission. This path adds:

- API for %CONEVAL AT command.
- API documentation.
- Helper function parse_coneval() used to parse parameters from a
  CONEVAL response.
- Test for parse_coneval().

Closes CIA-100

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-4318
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

simensrostad commented:
Depends on upcoming changes to lte link controller.

tokangas commented on lib/lte_link_control/lte_lc_helpers.c:
```diff
@@ -528,3 +528,160 @@ int parse_cereg(const char *at_response,
+
+	err = at_params_short_get(&resp_list, AT_CONEVAL_DL_PATHLOSS_INDEX, &params->dl_pathloss);
+	if (err) {
+		LOG_ERR("Could not parse download pathloss parameter, err: %d", err);
```
```suggestion
		LOG_ERR("Could not parse downlink pathloss parameter, err: %d", err);
```

simensrostad commented on include/modem/lte_lc.h:
```diff
@@ -204,9 +207,115 @@ enum lte_lc_pdp_type {
+};
+
+enum lte_lc_ce_level {
+	LTE_LC_CE_LEVEL_NO_REPETITION		= 0,
+	LTE_LC_CE_LEVEL_LOW_REPETITION		= 1,
+	LTE_LC_CE_LEVEL_MEDIUM_REPETITION	= 2,
+	LTE_LC_CE_LEVEL_HIGH_REPETITION		= 3,
```
Good idea. Ill add the CE level number in the enum name.

simensrostad commented on include/modem/lte_lc.h:
```diff
@@ -204,9 +207,115 @@ enum lte_lc_pdp_type {
+	 *
+	 * RRC CONNECTED: Currently configured CE level.
+	 *
+	 * CE levels 0-2 are supported in NB-IoT.
+	 * CE levels 0-1 are supported in LTE-M.
```
Lets just remove the enum. If questions are raised concerning this the AT command will always clarify.

simensrostad commented on include/modem/lte_lc.h:
```diff
@@ -204,9 +207,115 @@ enum lte_lc_pdp_type {
+	/* Reduction in power density in dB. */
+	int16_t dl_pathloss;
+
+	/* Current RSRP level at time of report. */
+	int16_t rsrp;
```
LTE LC does not currently do any conversion on such values. Given the nature of how these values are specified in the AT command doc I thought it may be best to leave it up to the application on they should interpreted/handled. That is also why they do not have units documented.

simensrostad replied:
If you find it sufficient I can document the ranges for the concerning parameters (rsrp, rsrq and snr) reflecting the AT command spec.

tokangas replied:
I'm not sure if we have had similar functionality in LTE LC earlier and how it has been implemented, but just though that it would be easier for the application developers if the values would already be converted to dB or dBm values. I guess we can thank 3GPP for the values being like this in the AT commands...

If we don't want to convert the values in LTE LC, then we should either describe the mapping or refer to the AT command spec here, I guess.

simensrostad commented on lib/at_cmd/Kconfig:
```diff
@@ -32,7 +32,7 @@ config AT_CMD_THREAD_PRIO
 config AT_CMD_THREAD_STACK_SIZE
 	int "AT thread stack size"
-	default 1024
+	default 2048
```
Oversaw that one. Forgot to trim the stack before pushing.

simensrostad commented on lib/lte_link_control/lte_lc.c:
```diff
@@ -1310,6 +1310,43 @@ int lte_lc_lte_mode_get(enum lte_lc_lte_mode *mode)
 	return 0;
 }
 
+int lte_lc_conn_eval_params_get(struct lte_lc_conn_eval_params *params)
```
Yes. A little blunder on my part there. How about changing the APIs to return positive ```<result>``` codes in the case of a failed evaluation? Adding the ```<result>``` code to the ```lte_lc_conn_eval_params``` structure would basically mean that you would have both valid and invalid members of the structure after the API call in the case of a failed evaluation which seems a little unorthodox and could be slightly confusing.

tokangas replied:
There could be an enum for possible evaluation results and that could be the first field in `lte_lc_conn_eval_params` struct with a note, that all other parameters in the struct are valid only when evaluation was successful.

But using values 0-7 to return the evaluation result is also fine for me.

simensrostad commented:
@tokangas Thank you for the review. I think I've addressed all your review points.

tokangas reviewed (approved):
👍

jtguggedal commented on include/modem/lte_lc.h:
```diff
@@ -204,9 +207,158 @@ enum lte_lc_pdp_type {
+enum lte_lc_tau_triggered {
+	LTE_LC_CELL_IN_TAI_LIST		= 0,
+
+	/* Cell not in TAI list, TAU will be triggered. */
```
The comments need to start with `/**` to be picked up by Doxygen. Applies to most new comments on struct members in the header file.

jtguggedal commented on include/modem/lte_lc.h:
```diff
@@ -204,9 +207,158 @@ enum lte_lc_pdp_type {
+	int16_t tx_power;
+
+	/* Four byte E-UTRAN cell ID in hexadecimal format. */
+	char cell_id[AT_CONEVAL_CELL_ID_STRING_LENGTH_MAX];
```
As discussed offline, these strings could be integers to align with how it's done elsewhere in the link controller

jtguggedal commented on include/modem/lte_lc.h:
```diff
@@ -204,9 +207,158 @@ enum lte_lc_pdp_type {
+	char cell_id[AT_CONEVAL_CELL_ID_STRING_LENGTH_MAX];
+
+	/* String of MCC and MNC values. */
+	char plmn[AT_CONEVAL_PLMN_STRING_LENGTH_MAX];
```
In #4356, I chose to split into MCC and MCC and present as integers. That could be considered here as well.

jtguggedal commented on include/modem/lte_lc.h:
```diff
@@ -479,6 +631,27 @@ int lte_lc_func_mode_get(enum lte_lc_func_mode *mode);
+ *
+ * @param params Pointer to structure to hold connection evaluation parameters.
+ *
+ * @return Zero on success or (negative) error code otherwise.
```
With the comments below explaining that return value can be a positive integer, I think this comment is a bit off.

jtguggedal commented on include/modem/lte_lc.h:
```diff
@@ -479,6 +631,27 @@ int lte_lc_func_mode_get(enum lte_lc_func_mode *mode);
+ * If the API call succeeds but connection evaluation fails due to modem/network related reasons,
+ * a positive error code is returned indicating the reason of failure.
+ *
+ * @return 1 Evaluation failed, no cell available.
```
I think `@retval` should be used instead of `@return` here.

jtguggedal commented on lib/at_cmd/Kconfig:
```diff
@@ -32,7 +32,7 @@ config AT_CMD_THREAD_PRIO
 config AT_CMD_THREAD_STACK_SIZE
 	int "AT thread stack size"
-	default 1024
+	default 1344
```
Perhaps keep the default and add `default 1344 if LTE_LINK_CONTROL`

jtguggedal commented on lib/lte_link_control/lte_lc.c:
```diff
@@ -1309,6 +1309,46 @@ int lte_lc_lte_mode_get(enum lte_lc_lte_mode *mode)
+	int err;
+	char buf[AT_CONEVAL_RESPONSE_MAX_LEN] = {0};
+	enum lte_lc_func_mode mode;
+
```
We need a NULL check for `params` up here somewhere

simensrostad commented:
@jtguggedal PR rebased on top of https://github.com/nrfconnect/sdk-nrf/pull/4356 and review comments addressed:
- PLMN is now stored as MCC and MNC integer values.
- Cell ID is also stored as an integer value.
- Slight cleanup in API docs explaining the various return values.
- Added conditional stack increase in ```AT_CMD``` kconfig.
- Updated tests to check for integer values rather than strings for MNC, MCC and cell ID.

tokangas commented on include/modem/lte_lc.h:
```diff
@@ -214,9 +317,161 @@ enum lte_lc_pdp_type {
+	 */
+	enum lte_lc_energy_estimate energy_estimate;
+
+	/** Is the evaluated cell part of the current TAI list. */
```
What's the TAI list? I think we should open this acronym. I'll try to figure out what it is...

tokangas replied:
TAI means Tracking Area Identifier. This list is got from the network and if a cell is on that list, no TAU is needed when switching to that cell.

simensrostad replied:
I was wondering the exact same thing. Thanks for the explanation, will update.

simensrostad replied:
Added additional description to both the variable and enums.

tokangas commented on include/modem/lte_lc.h:
```diff
@@ -214,9 +317,161 @@ enum lte_lc_pdp_type {
+
+	/** Coverage Enhancement level for PRACH depending on RRC state measured in.
+	 *
+	 *  RRC IDLE: CE level is estimated based on RSRP and network configueration.
```
```suggestion
	 *  RRC IDLE: CE level is estimated based on RSRP and network configuration.
```

rlubos commented:
Do we need `doc-required` label here? There're no `*.rst` files in the PR. Unless you've requested a tech-writer to check the headers.

simensrostad commented:
> Do we need `doc-required` label here? There're no `*.rst` files in the PR. Unless you've requested a tech-writer to check the headers.

The label was added for changes in the release notes. Looks like those changes was lost in a rebase, will re-add.
